File: modules/virtual-people-core-serving/src/main/kotlin/org/wfanet/virtualpeople/core/selector/VidModelSelector.kt
// Copyright 2023 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.selector

import java.nio.ByteOrder
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit
import kotlin.math.abs
import org.wfanet.measurement.api.v2alpha.ModelLine
import org.wfanet.measurement.api.v2alpha.ModelRollout
import org.wfanet.measurement.common.OpenEndTimeRange
import org.wfanet.measurement.common.toInstant
import org.wfanet.measurement.common.toLocalDate
import org.wfanet.measurement.common.toLong
import org.wfanet.virtualpeople.common.LabelerInput
import org.wfanet.virtualpeople.core.common.Hashing
import org.wfanet.virtualpeople.core.selector.resourcekey.ModelLineKey
import org.wfanet.virtualpeople.core.selector.resourcekey.ModelRolloutKey

const val CACHE_SIZE = 60
const val UPPER_BOUND_PERCENTAGE_ADOPTION = 1.1

/**
 * A utility class to determine what VID Model must be used to label events.
 *
 * Each ModelLine uses a different instance of this class. In case either the ModelLine or the list
 * of ModelRollouts change, a new instance must be created.
 *
 * @property modelLine the ModelLine used to determine what ModelRelease must be used to label
 *   events
 * @property rollouts the list of rollouts contained in the modelLine
 */
class VidModelSelector(private val modelLine: ModelLine, private val rollouts: List<ModelRollout>) {
  init {
    val modelLineId = ModelLineKey.fromName(modelLine.name)?.modelLineId
    require(modelLineId != null) { "ModelLine resource name is either unspecified or invalid" }
    for (rollout in rollouts) {
      val modelRolloutModelLineId = ModelRolloutKey.fromName(rollout.name)?.modelLineId
      require(modelLineId == modelRolloutModelLineId) {
        "ModelRollouts must be parented by the provided ModelLine."
      }
    }
  }

  /**
   * The adoption percentage of models is calculated each day. It consists of an array of
   * ModelReleasePercentile where each ModelReleasePercentile wraps the percentage of adoption of a
   * particular ModelRelease and the ModelRelease itself. LruCache keys are LocalDate objects
   * expressed in UTC time.
   *
   * Elements in the ArrayList are sorted by ModelRollout rollout start time from the oldest to the
   * most recent.
   */
  private val lruCache: LruCache<LocalDate, List<ModelReleasePercentile>> = LruCache(CACHE_SIZE)

  fun getModelRelease(labelerInput: LabelerInput): String? {
    val eventTimestampSec = labelerInput.timestampUsec / 1_000_000L
    val eventInstant = Instant.ofEpochSecond(eventTimestampSec)
    val modelLineActiveRange =
      OpenEndTimeRange(modelLine.activeStartTime.toInstant(), modelLine.activeEndTime.toInstant())
    if (eventInstant in modelLineActiveRange) {
      val eventDateUtc = eventInstant.atZone(ZoneOffset.UTC).toLocalDate()
      val modelAdoptionPercentages = readFromCache(eventDateUtc)
      var selectedModelRelease =
        if (!modelAdoptionPercentages.isEmpty()) {
          modelAdoptionPercentages.get(0).modelReleaseResourceKey
        } else {
          return null
        }
      val eventId = getEventId(labelerInput)
      for (percentage in modelAdoptionPercentages) {
        val eventFingerprint =
          Hashing.hashFingerprint64(
              buildString {
                append(percentage.modelReleaseResourceKey)
                append(eventId)
              }
            )
            .toLong(ByteOrder.LITTLE_ENDIAN)
        val reducedEventId = abs(eventFingerprint.toDouble() / Long.MAX_VALUE)
        if (reducedEventId < percentage.endPercentile) {
          selectedModelRelease = percentage.modelReleaseResourceKey
        }
      }
      return selectedModelRelease
    }
    return null
  }

  /**
   * Access to the cache is synchronized to prevent multiple threads calculating percentages in case
   * of cache miss.
   */
  private fun readFromCache(eventDateUtc: LocalDate): List<ModelReleasePercentile> {
    synchronized(this) {
      if (lruCache.containsKey(eventDateUtc)) {
        return lruCache[eventDateUtc]!!
      } else {
        val percentages = calculatePercentages(eventDateUtc)
        lruCache[eventDateUtc] = percentages
        return percentages
      }
    }
  }

  /**
   * Return a list of ModelReleasePercentile(s). Each ModelReleasePercentile wraps the percentage of
   * adoption of a particular ModelRelease and the ModelRelease itself. The list is sorted by either
   * rollout_period_start_date or instant_rollout_date.
   *
   * The adoption percentage of each ModelRollout is calculated as follows: (EVENT_DAY
   * - ROLLOUT_START_DAY) / (ROLLOUT_END_DAY - ROLLOUT_START_DAY).
   *
   * In case a ModelRollout has the `rollout_freeze_date` set and the event day is greater than
   * rollout_freeze_date, the EVENT_DAY in the above formula is replaced by `rollout_freeze_date` to
   * ensure that the rollout stops its expansion: (ROLLOUT_FREEZE_DATE - ROLLOUT_START_DAY) /
   * (ROLLOUT_END_DAY - ROLLOUT_START_DAY).
   *
   * In case of an instant rollout ROLLOUT_START_DATE is equal to ROLLOUT_END_DATE.
   */
  private fun calculatePercentages(eventDateUtc: LocalDate): List<ModelReleasePercentile> {
    val activeRollouts: List<ModelRollout> = retrieveActiveRollouts(eventDateUtc)
    return activeRollouts.map { activeRollout ->
      ModelReleasePercentile(
        calculatePercentageAdoption(eventDateUtc, activeRollout),
        activeRollout.modelRelease
      )
    }
  }

  /**
   * Returns the percentage of events that this ModelRollout must label for the given `eventDateUtc`
   */
  private fun calculatePercentageAdoption(
    eventDateUtc: LocalDate,
    modelRollout: ModelRollout
  ): Double {
    val modelRolloutFreezeDate =
      if (modelRollout.hasRolloutFreezeDate()) {
        modelRollout.rolloutFreezeDate.toLocalDate()
      } else {
        LocalDate.MAX
      }

    val rolloutPeriodStartDate =
      if (modelRollout.hasGradualRolloutPeriod()) {
        modelRollout.gradualRolloutPeriod.startDate.toLocalDate()
      } else {
        modelRollout.instantRolloutDate.toLocalDate()
      }
    val rolloutPeriodEndDate =
      if (modelRollout.hasGradualRolloutPeriod()) {
        modelRollout.gradualRolloutPeriod.endDate.toLocalDate()
      } else {
        modelRollout.instantRolloutDate.toLocalDate()
      }

    return if (rolloutPeriodStartDate == rolloutPeriodEndDate) {
      UPPER_BOUND_PERCENTAGE_ADOPTION
    } else if (eventDateUtc >= modelRolloutFreezeDate) {
      (ChronoUnit.DAYS.between(rolloutPeriodStartDate, modelRolloutFreezeDate).toDouble()) /
        (ChronoUnit.DAYS.between(rolloutPeriodStartDate, rolloutPeriodEndDate).toDouble())
    } else {
      (ChronoUnit.DAYS.between(rolloutPeriodStartDate, eventDateUtc).toDouble()) /
        (ChronoUnit.DAYS.between(rolloutPeriodStartDate, rolloutPeriodEndDate).toDouble())
    }
  }

  /**
   * Iterates through all available ModelRollout(s) sorted by rollout_period_start_time from the
   * most recent to the oldest. The function keeps adding ModelRollout(s) to the `activeRollouts`
   * array until the following condition is met: eventDay >= rolloutPeriodEndTime &&
   * !rollout.hasRolloutFreezeTime()
   */
  private fun retrieveActiveRollouts(eventDateUtc: LocalDate): List<ModelRollout> {
    if (rollouts.isEmpty()) {
      return rollouts
    }
    val sortedRollouts: List<ModelRollout> =
      rollouts.sortedBy {
        if (it.hasGradualRolloutPeriod()) {
          it.gradualRolloutPeriod.startDate.toLocalDate()
        } else {
          it.instantRolloutDate.toLocalDate()
        }
      }
    val firstRolloutPeriodStartDate =
      if (sortedRollouts.first().hasGradualRolloutPeriod()) {
        sortedRollouts.first().gradualRolloutPeriod.startDate
      } else {
        sortedRollouts.first().instantRolloutDate
      }
    if (eventDateUtc < firstRolloutPeriodStartDate.toLocalDate()) {
      return emptyList()
    }
    val activeRollouts = mutableListOf<ModelRollout>()
    for (rollout in sortedRollouts.asReversed()) {
      val rolloutPeriodEndDate =
        if (rollout.hasGradualRolloutPeriod()) {
          rollout.gradualRolloutPeriod.endDate.toLocalDate()
        } else {
          rollout.instantRolloutDate.toLocalDate()
        }
      if (eventDateUtc >= rolloutPeriodEndDate) {
        activeRollouts.add(rollout)
        if (!rollout.hasRolloutFreezeDate()) {
          // Stop only if there is no Freeze time set, otherwise other rollouts are taken until one
          // without rollout_freeze_time is found or no other rollouts are available
          break
        }
        continue
      }
      val rolloutPeriodStartDate =
        if (rollout.hasGradualRolloutPeriod()) {
          rollout.gradualRolloutPeriod.startDate.toLocalDate()
        } else {
          rollout.instantRolloutDate.toLocalDate()
        }
      if (eventDateUtc >= rolloutPeriodStartDate) {
        activeRollouts.add(rollout)
      }
    }
    return activeRollouts.asReversed()
  }

  private fun getEventId(labelerInput: LabelerInput): String {
    if (labelerInput.hasProfileInfo()) {
      val profileInfo = labelerInput.profileInfo
      if (profileInfo.emailUserInfo.hasUserId()) {
        return profileInfo.emailUserInfo.userId
      }
      if (profileInfo.phoneUserInfo.hasUserId()) {
        return profileInfo.phoneUserInfo.userId
      }
      if (profileInfo.loggedInIdUserInfo.hasUserId()) {
        return profileInfo.loggedInIdUserInfo.userId
      }
      if (profileInfo.loggedOutIdUserInfo.hasUserId()) {
        return profileInfo.loggedOutIdUserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace1UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace1UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace2UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace2UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace3UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace3UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace4UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace4UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace5UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace5UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace6UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace6UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace7UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace7UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace8UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace8UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace9UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace9UserInfo.userId
      }
      if (profileInfo.proprietaryIdSpace10UserInfo.hasUserId()) {
        return profileInfo.proprietaryIdSpace10UserInfo.userId
      }
    } else if (labelerInput.eventId.hasId()) {
      return labelerInput.eventId.id
    }
    error("Neither user_id nor event_id was found in the LabelerInput.")
  }

  private data class ModelReleasePercentile(
    val endPercentile: Double,
    val modelReleaseResourceKey: String,
  )
}

File: modules/virtual-people-core-serving/src/main/cc/wfa/virtual_people/core/selector/vid_model_selector.cc
// Copyright 2023 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/core/selector/vid_model_selector.h"

#include <google/protobuf/util/time_util.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <utility>

#include "absl/time/civil_time.h"
#include "absl/time/time.h"
#include "common_cpp/macros/macros.h"
#include "src/farmhash.h"

namespace wfa_virtual_people {

namespace {

const int kCacheSize = 60;
const double kUpperBoundPercentageAdoption = 1.1;

// Returns the model_line_id from the given resource name.
// Returns an empty string if not model_line_id is found.
std::string_view ReadModelLine(std::string_view input) {
  std::string_view modelLineMarker = "modelLines/";
  std::string_view modelRolloutMarker = "/modelRollouts/";

  size_t start = input.find(modelLineMarker);
  if (start != std::string_view::npos) {
    start += modelLineMarker.length();
    size_t end = input.find(modelRolloutMarker, start);
    if (end != std::string_view::npos) {
      return input.substr(start, end - start);
    } else {
      return input.substr(start);
    }
  }

  return "";
}

// Converts a given absl::Time into a absl::CivilDay object.
// absl::CivilDay is used to compare dates in UTC time.
absl::CivilDay TimeUsecToCivilDay(const absl::Time time) {
  absl::TimeZone utc_time_zone = absl::UTCTimeZone();
  return absl::ToCivilDay(time, utc_time_zone);
}

// Converts a google::type::Date object into a absl::CivilDay.
absl::CivilDay DateToCivilDay(const google::type::Date& date) {
  return absl::CivilDay(date.year(), date.month(), date.day());
}

double GetTimeDifferenceInSeconds(const absl::CivilDay& date_1,
                                  const absl::CivilDay& date_2) {
  absl::Time time_1 = absl::FromCivil(date_1, absl::UTCTimeZone());
  absl::Time time_2 = absl::FromCivil(date_2, absl::UTCTimeZone());
  absl::Duration difference = time_2 - time_1;
  return absl::ToDoubleSeconds(difference);
}

}  // namespace

VidModelSelector::VidModelSelector(
    const ModelLine& model_line,
    const std::vector<ModelRollout>& model_rollouts)
    : model_line_(model_line),
      model_rollouts_(model_rollouts),
      lru_cache_(kCacheSize) {}

// Move constructor
VidModelSelector::VidModelSelector(VidModelSelector&& other) noexcept
    : model_line_(std::move(other.model_line_)),
      model_rollouts_(std::move(other.model_rollouts_)),
      lru_cache_(std::move(other.lru_cache_)) {}

absl::StatusOr<VidModelSelector> VidModelSelector::Build(
    const ModelLine& model_line,
    const std::vector<ModelRollout>& model_rollouts) {
  std::string_view model_line_id = ReadModelLine(model_line.name());
  if (model_line_id == "") {
    return absl::InvalidArgumentError(
        "ModelLine resource name is either unspecified or invalid");
  }
  for (const ModelRollout& model_rollout : model_rollouts) {
    if (ReadModelLine(model_rollout.name()) != model_line_id) {
      return absl::InvalidArgumentError(
          "ModelRollouts must be parented by the provided ModelLine");
    }
  }

  VidModelSelector selector(model_line, model_rollouts);
  return std::move(selector);
}

absl::StatusOr<std::optional<std::string>> VidModelSelector::GetModelRelease(
    const LabelerInput& labeler_input) const {
  absl::Time event_timestamp =
      absl::FromUnixMicros(labeler_input.timestamp_usec());
  absl::Time model_line_active_start_time = absl::FromUnixMicros(
      google::protobuf::util::TimeUtil::TimestampToMicroseconds(
          model_line_.active_start_time()));

  absl::Time model_line_active_end_time =
      model_line_.has_active_end_time()
          ? absl::FromUnixMicros(
                google::protobuf::util::TimeUtil::TimestampToMicroseconds(
                    model_line_.active_end_time()))
          : absl::InfiniteFuture();

  if (event_timestamp >= model_line_active_start_time &&
      event_timestamp < model_line_active_end_time) {
    absl::CivilDay event_date_utc = TimeUsecToCivilDay(event_timestamp);
    std::vector<ModelReleasePercentile> model_adoption_percentages =
        ReadFromCache(event_date_utc);
    std::string selected_model_release;
    if (model_adoption_percentages.empty()) {
      return std::nullopt;
    } else {
      selected_model_release =
          model_adoption_percentages[0].model_release_resource_key;
    }
    std::string event_id;
    ASSIGN_OR_RETURN(event_id, GetEventId(labeler_input));

    for (auto percentage = model_adoption_percentages.begin();
         percentage != model_adoption_percentages.end(); ++percentage) {
      std::string string_to_hash;
      string_to_hash += percentage->model_release_resource_key;
      string_to_hash += event_id;
      // Convert into a signed number to make sure that the event_fingerprint is
      // the same one produced by the Kotlin library.
      int64_t event_fingerprint =
          static_cast<int64_t>(util::Fingerprint64(string_to_hash));

      double reduced_event_id =
          std::abs(static_cast<double>(event_fingerprint) /
                   std::numeric_limits<int64_t>::max());
      if (reduced_event_id < percentage->end_percentile) {
        selected_model_release = percentage->model_release_resource_key;
      }
    }
    return selected_model_release;
  } else {
    return std::nullopt;
  }
}

std::vector<ModelReleasePercentile> VidModelSelector::ReadFromCache(
    const absl::CivilDay& event_date_utc) const {
  std::lock_guard<std::mutex> lock(mtx_);

  std::optional<std::vector<ModelReleasePercentile>> cache_value =
      lru_cache_.Get(event_date_utc);
  if (cache_value.has_value()) {
    return cache_value.value();
  } else {
    std::vector<ModelReleasePercentile> percentages =
        CalculatePercentages(event_date_utc);
    lru_cache_.Add(event_date_utc, percentages);
    return percentages;
  }
}

std::vector<ModelReleasePercentile> VidModelSelector::CalculatePercentages(
    const absl::CivilDay& event_date_utc) const {
  std::vector<ModelRollout> active_rollouts =
      RetrieveActiveRollouts(event_date_utc);
  std::vector<ModelReleasePercentile> result;

  for (const ModelRollout& active_rollout : active_rollouts) {
    double percentage =
        CalculatePercentageAdoption(event_date_utc, active_rollout);
    result.emplace_back(percentage, active_rollout.model_release());
  }

  return result;
}

double VidModelSelector::CalculatePercentageAdoption(
    const absl::CivilDay& event_date_utc,
    const ModelRollout& model_rollout) const {
  absl::CivilDay model_rollout_freeze_date =
      model_rollout.has_rollout_freeze_date()
          ? DateToCivilDay(model_rollout.rollout_freeze_date())
          : absl::CivilDay(absl::CivilSecond::max());
  absl::CivilDay rollout_period_start_date =
      model_rollout.has_gradual_rollout_period()
          ? DateToCivilDay(model_rollout.gradual_rollout_period().start_date())
          : DateToCivilDay(model_rollout.instant_rollout_date());

  absl::CivilDay rollout_period_end_date =
      model_rollout.has_gradual_rollout_period()
          ? DateToCivilDay(model_rollout.gradual_rollout_period().end_date())
          : DateToCivilDay(model_rollout.instant_rollout_date());

  if (rollout_period_start_date == rollout_period_end_date) {
    return kUpperBoundPercentageAdoption;
  } else if (event_date_utc >= model_rollout_freeze_date) {
    return (GetTimeDifferenceInSeconds(rollout_period_start_date,
                                       model_rollout_freeze_date)) /
           (GetTimeDifferenceInSeconds(rollout_period_start_date,
                                       rollout_period_end_date));
  } else {
    return (GetTimeDifferenceInSeconds(rollout_period_start_date,
                                       event_date_utc)) /
           (GetTimeDifferenceInSeconds(rollout_period_start_date,
                                       rollout_period_end_date));
  }
}

bool VidModelSelector::CompareModelRollouts(const ModelRollout& lhs,
                                            const ModelRollout& rhs) const {
  const absl::CivilDay lhsDate =
      lhs.has_gradual_rollout_period()
          ? DateToCivilDay(lhs.gradual_rollout_period().start_date())
          : DateToCivilDay(lhs.instant_rollout_date());
  const absl::CivilDay rhsDate =
      rhs.has_gradual_rollout_period()
          ? DateToCivilDay(rhs.gradual_rollout_period().start_date())
          : DateToCivilDay(rhs.instant_rollout_date());
  return lhsDate < rhsDate;
}

std::vector<ModelRollout> VidModelSelector::RetrieveActiveRollouts(
    const absl::CivilDay& event_date_utc) const {
  if (model_rollouts_.empty()) {
    return model_rollouts_;
  }

  std::vector<ModelRollout> sorted_rollouts = model_rollouts_;
  std::sort(sorted_rollouts.begin(), sorted_rollouts.end(),
            [this](const ModelRollout& lhs, const ModelRollout& rhs) {
              return CompareModelRollouts(lhs, rhs);
            });

  const ModelRollout& first_rollout = sorted_rollouts.front();
  absl::CivilDay first_rollout_period_start_date =
      first_rollout.has_gradual_rollout_period()
          ? DateToCivilDay(first_rollout.gradual_rollout_period().start_date())
          : DateToCivilDay(first_rollout.instant_rollout_date());

  if (event_date_utc < first_rollout_period_start_date) {
    return {};
  }

  std::vector<ModelRollout> active_rollouts;

  for (auto rollout_idx = sorted_rollouts.rbegin();
       rollout_idx != sorted_rollouts.rend(); ++rollout_idx) {
    const ModelRollout& rollout = *rollout_idx;

    absl::CivilDay rollout_period_end_date =
        rollout.has_gradual_rollout_period()
            ? DateToCivilDay(rollout.gradual_rollout_period().end_date())
            : DateToCivilDay(rollout.instant_rollout_date());

    if (event_date_utc >= rollout_period_end_date) {
      active_rollouts.push_back(rollout);
      if (!rollout.has_rollout_freeze_date()) {
        break;
      }
      continue;
    }

    absl::CivilDay rollout_period_start_date =
        rollout.has_gradual_rollout_period()
            ? DateToCivilDay(rollout.gradual_rollout_period().start_date())
            : DateToCivilDay(rollout.instant_rollout_date());

    if (event_date_utc >= rollout_period_start_date) {
      active_rollouts.push_back(rollout);
    }
  }

  std::reverse(active_rollouts.begin(), active_rollouts.end());
  return active_rollouts;
}

absl::StatusOr<std::string> VidModelSelector::GetEventId(
    const LabelerInput& labeler_input) const {
  if (labeler_input.has_profile_info()) {
    const ProfileInfo* profile_info = &labeler_input.profile_info();

    if (profile_info->has_email_user_info() &&
        profile_info->email_user_info().has_user_id()) {
      return profile_info->email_user_info().user_id();
    }
    if (profile_info->has_phone_user_info() &&
        profile_info->phone_user_info().has_user_id()) {
      return profile_info->phone_user_info().user_id();
    }
    if (profile_info->has_logged_in_id_user_info() &&
        profile_info->logged_in_id_user_info().has_user_id()) {
      return profile_info->logged_in_id_user_info().user_id();
    }
    if (profile_info->has_logged_out_id_user_info() &&
        profile_info->logged_out_id_user_info().has_user_id()) {
      return profile_info->logged_out_id_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_1_user_info() &&
        profile_info->proprietary_id_space_1_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_1_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_2_user_info() &&
        profile_info->proprietary_id_space_2_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_2_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_3_user_info() &&
        profile_info->proprietary_id_space_3_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_3_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_4_user_info() &&
        profile_info->proprietary_id_space_4_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_4_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_5_user_info() &&
        profile_info->proprietary_id_space_5_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_5_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_6_user_info() &&
        profile_info->proprietary_id_space_6_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_6_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_7_user_info() &&
        profile_info->proprietary_id_space_7_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_7_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_8_user_info() &&
        profile_info->proprietary_id_space_8_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_8_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_9_user_info() &&
        profile_info->proprietary_id_space_9_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_9_user_info().user_id();
    }
    if (profile_info->has_proprietary_id_space_10_user_info() &&
        profile_info->proprietary_id_space_10_user_info().has_user_id()) {
      return profile_info->proprietary_id_space_10_user_info().user_id();
    }
  } else if (labeler_input.has_event_id()) {
    return labeler_input.event_id().id();
  }
  return absl::InvalidArgumentError(
      "Neither user_id nor event_id was found in the LabelerInput.");
}

}  // namespace wfa_virtual_people
